File: arrays/ltcd_661.py
'''
LeetCode - problem 661
'''

    def imageSmoother(self, img: List[List[int]]) -> List[List[int]]:
        m, n = len(img), len(img[0])
        res = [[0]*n for i in range(m)]
        
        dirs = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]
        
        for i in range(m):
            for j in range(n):
                cur_sum = img[i][j]
                count = 1
                for d in dirs:
                    x = i + d[0]
                    y = j + d[1]
                    if 0 <= x < m and 0 <= y < n:
                        count += 1
                        cur_sum += img[x][y]
                res[i][j] = int(cur_sum / count)
        
        return res

File: arrays/ltcd_661.cpp
//LeetCode - problem 661

    vector<vector<int>> imageSmoother(vector<vector<int>>& img) {
        int n = img.size();
        int m = img[0].size();
        vector<vector<int>> res = img;
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                float sum = img[i][j];
                int k = 1;
                
                if(i - 1 >= 0){
                    k++;
                    sum += img[i - 1][j];
                    if(j - 1 >= 0){
                        k++;
                        sum += img[i - 1][j - 1];
                    }
                    if(j + 1 < m){
                        k++;
                        sum += img[i - 1][j + 1];
                    }
                }
                if(i + 1 < n){
                    k++;
                    sum += img[i + 1][j];
                    if(j - 1 >= 0){
                        k++;
                        sum += img[i + 1][j - 1];
                    }
                    if(j + 1 < m){
                        k++;
                        sum += img[i + 1][j + 1];
                    }
                }
                if(j - 1 >= 0){
                    k++;
                    sum += img[i][j - 1];
                }
                if(j + 1 < m){
                    k++;
                    sum += img[i][j + 1];
                }
                res[i][j] = (int)(sum / k);
            }
        }
        return res;
    }
